File: _archived/notifications-nodejs/src/services/delivery-issue.service.ts
import { DeliveryIssueRepository } from "../repositories/delivery-issue.repository";
import { DeliveryIssueType } from "../models/delivery-issue.model";
import { NormalizedEventType } from "../models/notification-event.model";
import { NotificationChannel } from "../types";
import { auditClient } from "../integrations/audit/audit.client";
import { logger } from "../shared/logger";

interface DeliveryIssueContext {
  tenantId: string;
  notificationId: string;
  notificationAttemptId?: string | null;
  channel: NotificationChannel;
  provider: string;
  normalizedEventType: NormalizedEventType;
  rawEventType: string;
  recipientContact?: string | null;
  errorCode?: string | null;
  errorMessage?: string | null;
}

function buildRecommendedAction(
  issueType: DeliveryIssueType,
  channel: NotificationChannel
): string | null {
  if (issueType === "bounced_email" || issueType === "invalid_email") {
    return "Verify and update recipient email address. Consider retrying via SMS if phone number is available.";
  }
  if (issueType === "sms_undelivered" || issueType === "invalid_phone") {
    return "Verify recipient phone number. Consider retrying via email if address is available.";
  }
  if (issueType === "unsubscribed_recipient") {
    return "Recipient has opted out. Do not retry on this channel.";
  }
  if (issueType === "complained_recipient") {
    return "Recipient marked as spam. Suppress contact on this channel.";
  }
  if (issueType === "opted_out_recipient") {
    return "Recipient has opted out of SMS. Do not retry on this channel.";
  }
  if (issueType === "provider_rejected") {
    return "Review provider rejection reason and validate message content.";
  }
  return null;
}

export class DeliveryIssueService {
  private repo: DeliveryIssueRepository;

  constructor() {
    this.repo = new DeliveryIssueRepository();
  }

  async processEvent(ctx: DeliveryIssueContext): Promise<void> {
    const {
      tenantId,
      notificationId,
      notificationAttemptId,
      channel,
      provider,
      normalizedEventType,
      rawEventType,
      recipientContact,
      errorCode,
      errorMessage,
    } = ctx;

    let issueType: DeliveryIssueType | null = null;

    if (normalizedEventType === "bounced") {
      issueType = channel === "email" ? "bounced_email" : "sms_undelivered";
    } else if (normalizedEventType === "undeliverable") {
      issueType = channel === "sms" ? "sms_undelivered" : "provider_rejected";
    } else if (normalizedEventType === "rejected") {
      issueType = "provider_rejected";
    } else if (normalizedEventType === "complained") {
      issueType = "complained_recipient";
    } else if (normalizedEventType === "unsubscribed") {
      issueType = "unsubscribed_recipient";
    }

    if (!issueType) return;

    const recommendedAction = buildRecommendedAction(issueType, channel);
    const details = {
      rawEventType,
      normalizedEventType,
      recipientContact: recipientContact ?? undefined,
      errorCode: errorCode ?? undefined,
      errorMessage: errorMessage ?? undefined,
    };

    try {
      const issue = await this.repo.createIfNotExists({
        tenantId,
        notificationId,
        notificationAttemptId: notificationAttemptId ?? null,
        channel,
        provider,
        issueType,
        recommendedAction,
        detailsJson: JSON.stringify(details),
      });

      if (issue) {
        logger.info("Delivery issue created", {
          tenantId,
          notificationId,
          issueType,
          channel,
          provider,
        });

        await auditClient.publishEvent({
          eventType: "delivery_issue.created",
          tenantId,
          channel,
          provider,
          metadata: { notificationId, issueType, recommendedAction: recommendedAction ?? undefined },
        });
      }
    } catch (err) {
      logger.error("Failed to create delivery issue", {
        error: String(err),
        tenantId,
        notificationId,
        issueType,
      });
    }
  }
}

File: apps/services/notifications/Notifications.Infrastructure/Services/DeliveryIssueService.cs
using Microsoft.Extensions.Logging;
using Notifications.Application.Interfaces;
using Notifications.Domain;
using LegalSynq.AuditClient;
using LegalSynq.AuditClient.DTOs;

namespace Notifications.Infrastructure.Services;

public class DeliveryIssueServiceImpl : IDeliveryIssueService
{
    private readonly IDeliveryIssueRepository _repo;
    private readonly IAuditEventClient _auditClient;
    private readonly ILogger<DeliveryIssueServiceImpl> _logger;

    public DeliveryIssueServiceImpl(IDeliveryIssueRepository repo, IAuditEventClient auditClient, ILogger<DeliveryIssueServiceImpl> logger)
    {
        _repo = repo;
        _auditClient = auditClient;
        _logger = logger;
    }

    public async Task ProcessEventAsync(DeliveryIssueContext ctx)
    {
        string? issueType = ctx.NormalizedEventType switch
        {
            "bounced" => ctx.Channel == "email" ? "bounced_email" : "sms_undelivered",
            "undeliverable" => ctx.Channel == "sms" ? "sms_undelivered" : "provider_rejected",
            "rejected" => "provider_rejected",
            "complained" => "complained_recipient",
            "unsubscribed" => "unsubscribed_recipient",
            _ => null
        };

        if (issueType == null) return;

        var recommendedAction = GetRecommendedAction(issueType, ctx.Channel);
        var details = System.Text.Json.JsonSerializer.Serialize(new
        {
            rawEventType = ctx.RawEventType,
            normalizedEventType = ctx.NormalizedEventType,
            recipientContact = ctx.RecipientContact,
            errorCode = ctx.ErrorCode,
            errorMessage = ctx.ErrorMessage
        });

        try
        {
            var issue = await _repo.CreateIfNotExistsAsync(new DeliveryIssue
            {
                TenantId = ctx.TenantId,
                NotificationId = ctx.NotificationId,
                NotificationAttemptId = ctx.NotificationAttemptId,
                Channel = ctx.Channel,
                Provider = ctx.Provider,
                IssueType = issueType,
                RecommendedAction = recommendedAction,
                DetailsJson = details
            });

            if (issue != null)
            {
                _logger.LogInformation("Delivery issue created: {TenantId} {NotificationId} {IssueType}", ctx.TenantId, ctx.NotificationId, issueType);
                try { await _auditClient.IngestAsync(new IngestAuditEventRequest { EventType = "delivery_issue.created", Action = "delivery_issue.created", SourceSystem = "notifications", Description = $"Delivery issue created: {issueType}", Scope = new AuditEventScopeDto { TenantId = ctx.TenantId.ToString() } }); } catch { }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to create delivery issue: {TenantId} {NotificationId} {IssueType}", ctx.TenantId, ctx.NotificationId, issueType);
        }
    }

    private static string? GetRecommendedAction(string issueType, string channel) => issueType switch
    {
        "bounced_email" or "invalid_email" => "Verify and update recipient email address. Consider retrying via SMS if phone number is available.",
        "sms_undelivered" or "invalid_phone" => "Verify recipient phone number. Consider retrying via email if address is available.",
        "unsubscribed_recipient" => "Recipient has opted out. Do not retry on this channel.",
        "complained_recipient" => "Recipient marked as spam. Suppress contact on this channel.",
        "opted_out_recipient" => "Recipient has opted out of SMS. Do not retry on this channel.",
        "provider_rejected" => "Review provider rejection reason and validate message content.",
        _ => null
    };
}
